### PR TITLE
Update Cypress.log to throw when not provided an obj + handle being called outside of command

### LINKED ISSUE #3658: Throws error on Cypress.log without previous command
Cypress 3.1.5 on Mac

```js
/// <reference types="cypress" />
it('works', () => {
  Cypress.log('after hook')
})
```

Throws error 
screen shot 2019-03-06 at 12 05 52 pm

Code that throws

```js
    _.defaults(obj, current != null ? current.pick("name", "type") : void 0);
    if (!parentOrChildRe.test(obj.type)) {
      obj.type = current.hasPreviouslyLinkedCommand() ? "child" : "parent";
    }
```

https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cypress/log.coffee#L117

### PR DESCRIPTION
**Pull request by jennifer-shehane**
- close #3658

### CODE REVIEW

**Comment by Bkucera** [comment]
@jennifer-shehane are you sure this fixes #3658 ? That issue is referring to calling Cypress.log(...) before a cypress command

**Comment by vrockai** [comment]
@Bkucera please see https://github.com/cypress-io/cypress/issues/3658#issuecomment-490274172
I've tried to come up with a solution, but it's not as trivial as adding a guard, new errors would pop up. At the end this doesn't really look like a bug to me.

**Comment by jennifer-shehane** [comment]
@Bkucera Yeah, you're right. This only partially addresses the issue of the args being passed incorrectly. 

I've added a failing snapshot test for the issue of this failing when called with correct args here: https://github.com/cypress-io/cypress/blob/issue-3658/packages/server/test/support/fixtures/projects/e2e/cypress/integration/issue_3658_spec.js

File: packages/driver/src/cypress/log.coffee [hunk 1]
```diff
@@ -490,13 +493,17 @@ create = (Cypress, cy, state, config) ->
     if _.isFunction(onBeforeLog)
       return if onBeforeLog.call(cy, log) is false
 
-    ## set the log on the command
+    ## set the log on the command
```
Comment (Bkucera) [inline packages/driver/src/cypress/log.coffee]: extra whitespace

File: packages/server/test/e2e/3_issue_3658_spec.coffee [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  e2e.setup()
+
+  ## https://github.com/cypress-io/cypress/issues/3658
+
+  it "passes", ->
+    e2e.exec(@, {
+      spec: "issue_3658_spec.js"
+      snapshot: true
+      expectedExitCode: 0
+    })
```
Comment (Bkucera) [inline packages/server/test/e2e/3_issue_3658_spec.coffee]: Does this have to be an e2e test? I don't see any output in the snapshot that's necessary to prove this test is valid. I think we could move this into an test in the driver

**Review by kuceb — CHANGES_REQUESTED** [review]

File: packages/server/test/e2e/3_issue_3658_spec.coffee (hunk 2 above)
Comment (jennifer-shehane) [reply packages/server/test/e2e/3_issue_3658_spec.coffee]: hmm, perhaps

Comment (Bkucera) [reply packages/server/test/e2e/3_issue_3658_spec.coffee]: That would be nice since it would run in <1s vs 30 seconds

**Review by kuceb — DISMISSED** [review]

**Pull request merged**